### PR TITLE
Mixed Poisson Process

### PR DESCRIPTION
**Pull request by Gabinou**
Hello,

As suggested, I have removed general messyness and included proposed feature into independent branches, starting with the simplest: the mixed poisson process.

Is this branch more suitable?

### CODE REVIEW

**Comment by crflynn** [comment]
See my comment in #7 

Looks like you have deleted some test fixtures? That's why the tests are failing on travis.

File: .travis.yml [hunk 1]
```diff
@@ -9,6 +9,7 @@ python:
 install:
   - pip install numpy scipy
   - pip install pytest pytest-cov pytest-xdist pytest-mock
+  - pip install pytest --upgrade
```
Comment (crflynn) [inline .travis.yml]: This line is not necessary. Can you remove it?

File: stochastic/continuous/__init__.py [hunk 2]
```diff
@@ -11,6 +11,7 @@
 from stochastic.continuous.inverse_gaussian import InverseGaussianProcess
 from stochastic.continuous.multifractional_brownian_motion import MultifractionalBrownianMotion
 from stochastic.continuous.poisson import PoissonProcess
+from stochastic.continuous.mixed_poisson import MixedPoissonProcess
```
Comment (crflynn) [inline stochastic/continuous/__init__.py]: Keep imports alphabetized here.

File: docs/continuous.rst [hunk 3]
```diff
@@ -16,6 +16,7 @@ The :py:mod:`stochastic.continuous` module provides classes for generating
 * :py:class:`stochastic.continuous.InverseGaussianProcess`
 * :py:class:`stochastic.continuous.MultifractionalBrownianMotion`
 * :py:class:`stochastic.continuous.PoissonProcess`
+* :py:class:`stochastic.continuous.MixedPoissonProcess`
```
Comment (crflynn) [inline docs/continuous.rst]: Keep this alphabetized.

File: stochastic/continuous/mixed_poisson.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    A Poisson process with rate :math:`\lambda` is a count of occurrences of
+    i.i.d. exponential random variables with mean :math:`1/\lambda`. This class
+    generates samples of times for which cumulative exponential random variables occur. 
+
+    :param function rate_func: random distribution of the rate :math:`\lambda`
+    :param rate_args: arguments to input into the rate_func function
+    :param rate_kwargs: keyword arguments to input into the rate_func function
+    """
+
+    def __init__(self, rate_func,*rate_args,**rate_kwargs):
```
Comment (crflynn) [inline stochastic/continuous/mixed_poisson.py]: Space after commas.

rate_args and rate_kwargs should be passed as single objects, not unpacked.

File: stochastic/continuous/mixed_poisson.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    def rate(self):
+        """Current rate."""
+        return self._rate
+        
+    @rate.setter
+    def rate(self, value):
+        rate_func, rate_args,rate_kwargs = value 
+        self._rate_func=rate_func
+        self._rate_args=rate_args
+        self._rate_kwargs=rate_kwargs
```
Comment (crflynn) [inline stochastic/continuous/mixed_poisson.py]: The rate setter and getter here is not needed. On sample you should just be doing self.rate = self.gen_rate() and gen_rate should return the new sample. Just use the parent rate getter/setter.

File: tests/continuous/conftest.py [hunk 6]
```diff
@@ -94,24 +94,6 @@ def volatility(request):
...
-    return 1
-
-@pytest.fixture(params=[mean_func_monotonic, None])
-def mean_func(request):
-    return request.param
-
-@pytest.fixture(params=[mean_func_not_monotonic, mean_func_no_args, 1])
-def mean_func_invalid(request):
-    return request.param
-
```
Comment (crflynn) [inline tests/continuous/conftest.py]: Please don't delete test fixtures for other processes.

File: tests/continuous/test_mixed_poisson.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+            s = instance.sample(n_fixture, length, zero)
+    elif length is not None and n_fixture is None:
+        s = instance.sample(n_fixture, length, zero)
+        assert s[-1] >= length
+    else:  # n_fixture is not None:
+        s = instance.sample(n_fixture, length, zero)
+        assert len(s) == n_fixture + int(zero)
+
+def test_poisson_process_times(rate_func,rate_args,rate_kwargs, n):
+    instance = MixedPoissonProcess(rate_func,*rate_args,**rate_kwargs)
```
Comment (crflynn) [inline tests/continuous/test_mixed_poisson.py]: Space after comma for this entire file.

File: stochastic/continuous/mixed_poisson.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        """Generate a new rate. Called when any parameter is set, and upon each call for samples."""
+        if (hasattr(self,'_rate_args')) & (hasattr(self,'_rate_func')) & (hasattr(self,'_rate_kwargs')) : 
+            self._rate = self._rate_func(*self._rate_args,**self.rate_kwargs)
+            self._check_nonnegative_number(self._rate, "Arrival rate")
+
+    def sample(self, n=None, length=None, zero=True):
+        out=super(MixedPoissonProcess,self).sample(n, length, zero)
+        self.gen_rate()
+        """Generate a new random rate upon each realization."""
+        return out
```
Comment (crflynn) [inline stochastic/continuous/mixed_poisson.py]: Docstring should be immediately after function signature.

Shouldn't the steps here be to 
* sample the rate
* call ``_sample_poisson_process(n, length, zero)``
?

File: docs/continuous.rst [hunk 9]
```diff
@@ -56,6 +57,9 @@ The :py:mod:`stochastic.continuous` module provides classes for generating
 
 .. autoclass:: stochastic.continuous.PoissonProcess
     :members: rate, sample
+    
+.. autoclass:: stochastic.continuous.MixedPoissonProcess
+    :members: rate_func,rate_args,rate_kwargs,gen_rate
```
Comment (crflynn) [inline docs/continuous.rst]: Should only expose the attributes/methods that should be public, here we really only want rate, rate_func, rate_args, rate_kwargs, sample.

File: stochastic/continuous/mixed_poisson.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        
+    @rate.setter
+    def rate(self, value):
+        rate_func, rate_args,rate_kwargs = value 
+        self._rate_func=rate_func
+        self._rate_args=rate_args
+        self._rate_kwargs=rate_kwargs
+        self.gen_rate()
+        
+    def gen_rate(self):
```
Comment (crflynn) [inline stochastic/continuous/mixed_poisson.py]: Make this function "private", i.e. ``_gen_rate()``

File: stochastic/continuous/mixed_poisson.py (hunk 8 above)
Comment (Gabinou) [reply stochastic/continuous/mixed_poisson.py]: No, it should not. That would mean outputting a realization for a rate different from what instance.rate shows.  _gen_rate() should be called after _sample_poisson_process().

File: stochastic/continuous/mixed_poisson.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
+"""Mixed poisson processes."""
+
+from stochastic.base import Checks
+from stochastic.continuous import PoissonProcess
```
Comment (crflynn) [inline stochastic/continuous/mixed_poisson.py]: Change this to 
`from stochastic.continuous.poisson import PoissonProcess`

This should eliminate the circular import that appears to be causing the tests to fail.

**Comment by Gabinou** [comment]
Indeed I couldn't... but now I can! The tests are passing now.

**Comment by Gabinou** [comment]
Indeed I couldn't... but now I can!

**Pull request merged**